### PR TITLE
GRPO: compute rewards over eval set

### PR DESCRIPTION
**Pull request by sidnarayanan**
Basically what the title says. Trainer's eval loop is built around losses/logits, so I am abusing the return and call signatures of `prediction_step` and `compute_metrics` respectively. This lets us compute+log rewards without rewriting the whole eval loop and logging mechanism. 

I also pulled in the changes from https://github.com/huggingface/trl/pull/2776, which are needed to unwrap the model for generation in the eval loop.

Most of the code diff is me pulling code out of `compute_loss` into separate methods, so they can be reused in `prediction_step`

### CODE REVIEW

File: trl/trainer/grpo_config.py [hunk 1]
```diff
@@ -174,27 +174,3 @@ class GRPOConfig(TrainingArguments):
...
-            "`π_ref = α * π_θ + (1 - α) * π_ref_prev`. To use this parameter, you must set `sync_ref_model=True`."
-        },
-    )
-    ref_model_sync_steps: int = field(
-        default=64,
-        metadata={
-            "help": "τ parameter from the TR-DPO paper, which determines how frequently the current policy is "
-            "synchronized with the reference policy. To use this parameter, you must set `sync_ref_model=True`."
-        },
-    )
```
Comment (sidnarayanan) [inline trl/trainer/grpo_config.py:200]: This doesn't work for large models, and checkpointing the best epochs (which eval metrics enables) lets us do the reset manually. So I'm removing the configuration.

**Review by jamesbraza — APPROVED** [review]
Approving

I am wondering, do you run their unit tests, or does this break them?

**Comment by sidnarayanan** [comment]
> I am wondering, do you run their unit tests, or does this break them?
I've not been, but I don't think this will - they only test GRPO eval in one place, and don't actually check the outputs - just that it runs.
